manual/src/tutorials/memorymodel.etex

```diff
@@ -0,0 +1,783 @@
+conflicting accesses, there is a data race.
+
+Both of the programs that we had seen in the section~\ref{s:why_relaxed_memory}
+have data races. It is no surprise that they exhibit non sequentially
+consistent behaviours.
+
+Accessing disjoint array indices and fields of a record in parallel is not a
+data race. For example,
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let a = [| 0; 1 |]
+let d1 () = a.(0) <- 42
+let d2 () = a.(1) <- 42
+\end{caml_example*}
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+type t = {
+  mutable a : int;
+  mutable b : int
+}
+let r = {a = 0; b = 1}
+let d1 () = r.a <- 42
+let d2 () = r.b <- 42
+\end{caml_example*}
+
+\noindent do not have data races.
+
+Races on atomic locations do not lead to a data race.
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let r = Atomic.make 0
+let d1 () = Atomic.set r 1
+let d2 () = Atomic.get r
+\end{caml_example*}
+
+\subsubsection{s:mm_msg_passing}{Message-passing}
+
+Atomic variables may be used for implementing non-blocking communication
+between domains.
+
+\begin{caml_example*}{verbatim}
+(* No data race *)
+let msg = ref 0
+let flag = Atomic.make false
+let d1 () =
+  msg := 42; (* a *)
+  Atomic.set flag true (* b *)
+let d2 () =
+  if Atomic.get flag (* c *) then
+    !msg (* d *)
+  else 0
+\end{caml_example*}
+
+Observe that the actions "a" and "d" write and read from the same non-atomic
+location "msg", respectively, and hence are conflicting. We need to establish
+that "a" and "d" have a happens-before relationship in order to show that this
+program does not have a data race.
+
+The action "a" precedes "b" in program order, and hence, "a" happens-before
+"b". Similarly, "c" happens-before "d". If "d2" observes the atomic variable
+"flag" to be "true", then "b" precedes "c" in happens-before order. Since
+happens-before is transitive, the conflicting actions "a" and "d" are in
+happens-before order. If "d2" observes the "flag" to be "false", then the read
+of "msg" is not done. Hence, there is no conflicting access in this execution
+trace. Hence, the program does not have a data race.
+
+The following modified version of the message passing program does have a data
+race.
+
+\begin{caml_example*}{verbatim}
+(* Has data race *)
+let msg = ref 0
+let flag = Atomic.make false
+let d1 () =
+  msg := 42; (* a *)
+  Atomic.set flag true (* b *)
+let d2 () =
+  ignore (Atomic.get flag); (* c *)
+  !msg (* d *)
+\end{caml_example*}
+
+The domain "d2" now unconditionally reads the non-atomic reference "msg".
+Consider the execution trace:
+
+\begin{verbatim}
+Atomic.get flag; (* c *)
+!msg; (* d *)
+msg := 42; (* a *)
+Atomic.set flag true (* b *)
+\end{verbatim}
+
+In this trace, "d" and "a" are conflicting operations. But there is no
+happens-before relationship between them. Hence, this program has a data race.
+
+\section{s:local_drf}{Local data race freedom}
+
+The OCaml memory model offers strong guarantees even for programs with data races.
```

Comment:
> Out of curiosity I searched for "safety" (as in memory safety) in this file and it returned nothing. The "easy bits" mention the absence of "crashes", so I wonder if it would not be a good idea to link up the informal claim with a more formal claim in a sentence or two in this section.

Let me add a sentence about memory safety.

> Also, constraints on the memory representation play a role, such as the tag being immutable from now on (and other things?). (This kind of details could be more important for people who are already interested by the memory representation, such as those doing things on the C side.)

Would be good to include this in this page. Should we aim to do this as a follow-up PR? This PR is already large and has been up for a while now.